File: src/utils/data/ListUtil.java
package utils.data;

import utils.math.MathUtil;

import java.util.*;

/**
 * Created with IntelliJ IDEA.
 * User: Tony Tsang
 * Date: 2012-11-07
 * Time: 09:37
 */
public class ListUtil {
    private ListUtil() {}

    public static <T extends Comparable<T>> ListDiff<T> compareListWithoutSorting(
            List<T> list, List<T> compareList
    ) {
        return compareListWithoutSorting(list, compareList, Comparable::compareTo);
    }

    public static <T extends Comparable<T>> ListDiff<T> compareList(
            List<T> list, List<T> compareList
    ) {
        return compareList(list, compareList, Comparable::compareTo);
    }

    public static <T extends Comparable<T>> ListDiff<T> compareSortedList(
            List<T> list, List<T> compareList
    ) {
        return compareSortedList(list, compareList, Comparable::compareTo);
    }

    public static <T> ListDiff<T> compareListWithoutSorting(
        List<T> list, List<T> compareList, Comparator<T> comparator
    ) {
        List<T> removedElements = new ArrayList<>(list), newElements = new ArrayList<>(compareList),
                sameElements = new ArrayList<>(list.size());

        Iterator<T> iter = removedElements.iterator(), cIter;

        while (iter.hasNext()) {
            T ele = iter.next();

            cIter = newElements.iterator();

            while (cIter.hasNext()) {
                T cEle = cIter.next();
                if (comparator.compare(ele, cEle) == 0) {
                    sameElements.add(ele);
                    cIter.remove();
                    iter.remove();
                    break;
                }
            }
        }

        return new ListDiff<>(newElements, removedElements, sameElements);
    }

    public static <T> ListDiff<T> compareList(
            List<T> list, List<T> compareList, Comparator<T> comparator
    ) {
        List<T> newList = new ArrayList<>(list), newCompareList = new ArrayList<>(compareList);
        Collections.sort(newList, comparator);
        Collections.sort(newCompareList, comparator);

        return compareSortedList(newList, newCompareList, comparator);
    }

    public static <T> ListDiff<T> compareSortedList(
            List<T> list, List<T> compareList, Comparator<T> comparator
    ) {
        List<T> sameElements = new ArrayList<>(list.size()),
                newElements = new ArrayList<>(compareList),
                removedElements = new ArrayList<>(list);

        if (removedElements.isEmpty() || newElements.isEmpty()) {
            return new ListDiff<>(sameElements, newElements, removedElements);
        }

        Iterator<T> iter = removedElements.iterator(), cIter = newElements.iterator();

        T ele = iter.next(), cele = cIter.next();

        while (ele != null && cele != null) {
            int cmp = comparator.compare(ele, cele);

            if (cmp > 0) {
                if (cIter.hasNext()) {
                    cele = cIter.next();
                } else {
                    break;
                }
            } else if (cmp < 0) {
                if (iter.hasNext()) {
                    ele = iter.next();
                } else {
                    break;
                }
            } else {
                sameElements.add(ele);
                iter.remove();
                cIter.remove();
                if (iter.hasNext()) {
                    ele = iter.next();
                } else {
                    break;
                }
                if (cIter.hasNext()) {
                    cele = cIter.next();
                } else {
                    break;
                }
            }
        }

        return new ListDiff<>(newElements, removedElements, sameElements);
    }

    public static <T extends Comparable<T>> boolean contains(List<T> list, List<T> compareList) {
        ListDiff<T> diff = compareList(list, compareList);

        return !diff.hasNewElements();
    }

    public static <T extends Comparable<T>> boolean intercept(List<T> list, List<T> compareList) {
        ListDiff<T> diff = compareList(list, compareList);

        return diff.hasSameElements();
    }

    public static <T extends Comparable<T>> boolean isListEquals(List<T> list, List<T> compareList) {
        return isListEquals(list, compareList, Comparable::compareTo);
    }

    public static <T extends Comparable<T>> boolean isListExactlyEquals(List<T> list, List<T> compareList) {
        return isListExactlyEquals(list, compareList, Comparable::compareTo);
    }

    public static <T> boolean contains(List<T> list, List<T> compareList, Comparator<T> comparator) {
        ListDiff<T> diff = compareList(list, compareList, comparator);

        return !diff.hasNewElements();
    }

    public static <T> boolean intercept(List<T> list, List<T> compareList, Comparator<T> comparator) {
        ListDiff<T> diff = compareList(list, compareList, comparator);

        return diff.hasSameElements();
    }

    public static <T> boolean isListEquals(List<T> list, List<T> compareList, Comparator<T> comparator) {
        if (list.size() != compareList.size()) {
            return false;
        } else {
            List<T> newList = new ArrayList<>(list), newCompareList = new ArrayList<>(compareList);

            Collections.sort(newList, comparator);
            Collections.sort(newCompareList, comparator);

            for (int i = 0; i < newList.size(); i++) {
                T ele = newList.get(i), cele = newCompareList.get(i);

                if (comparator.compare(ele, cele) != 0) {
                    return false;
                }
            }

            return true;
        }
    }

    public static <T> boolean isListExactlyEquals(List<T> list, List<T> compareList, Comparator<T> comparator) {
        if (list.size() != compareList.size()) {
            return false;
        } else {
            for (int i = 0; i < list.size(); i++) {
                T e1 = list.get(i), e2 = compareList.get(i);

                if (comparator.compare(e1, e2) != 0) {
                    return false;
                }
            }

            return true;
        }
    }

    public static <T> void reorderList(List<T> list, final List<T> refList) {
        Collections.sort(
            list, (o1, o2) -> {
                int i1 = refList.indexOf(o1), i2 = refList.indexOf(o2);
                if (i1 < i2) {
                    return -1;
                } else if (i1 > i2) {
                    return 1;
                } else {
                    return 0;
                }
            }
        );
    }

    public static <T> boolean syncList(List<T> targetList, ListDiff<T> listDiff) {
        boolean modified = false;

        if (listDiff.hasRemovedElements()) {
            targetList.removeAll(listDiff.getRemovedElements());
            modified = true;
        }

        if (listDiff.hasNewElements()) {
            targetList.addAll(listDiff.getNewElements());
            modified = true;
        }

        return modified;
    }

    public static <T> boolean syncLists(List<T> targetList, List<T> sourceList, Comparator<T> comparator) {
        return syncList(targetList, compareList(targetList, sourceList, comparator));
    }

    public static class ListDiff<T> {
        private List<T> removedElements, newElements, sameElements;

        public ListDiff(
            List<T> newElements, List<T> removedElements, List<T> sameElements
        ) {
            this.newElements = newElements;
            this.removedElements = removedElements;
            this.sameElements = sameElements;
        }

        public List<T> getRemovedElements() {
            return removedElements;
        }

        public List<T> getNewElements() {
            return newElements;
        }

        public List<T> getSameElements() {
            return sameElements;
        }

        public boolean hasRemovedElements() {
            return removedElements.size() > 0;
        }

        public boolean hasNewElements() {
            return newElements.size() > 0;
        }

        public boolean hasSameElements() {
            return sameElements.size() > 0;
        }

        public boolean isSame() {
            return newElements.isEmpty() && removedElements.isEmpty();
        }

        public boolean isNotSame() {
            return !isSame();
        }

        public int getNewElementCount() {
            return newElements.size();
        }

        public int getRemovedElementCount() {
            return removedElements.size();
        }

        public int getSameElementCount() {
            return sameElements.size();
        }
    }

    private static List<Integer> createList(int size) {
        List<Integer> list = new ArrayList<>(size);

        for (int i = 0; i < size; i++) {
            list.add(MathUtil.randomInteger(0, size));
        }

        return list;
    }
}

File: src/utils/data/ListUtil.kt
package utils.data

import utils.math.MathUtil
import java.util.*

/**
 * Created with IntelliJ IDEA.
 * User: Tony Tsang
 * Date: 2017-07-05
 * Time: 17:02
 */
class ListUtil private constructor() {
    companion object {
        @JvmStatic
        fun <T : Comparable<T>> compareListWithoutSorting(
            list: List<T>, compareList: List<T>
        ): ListDiff<T> {
            return compareListWithoutSorting(
                list, compareList, kotlin.Comparator { o1, o2 -> o1.compareTo(o2) }
            )
        }

        @JvmStatic
        fun <T : Comparable<T>> compareList(
            list: List<T>, compareList: List<T>
        ):ListDiff<T> {
            return compareList(list, compareList, Comparator { o1, o2 -> o1.compareTo(o2) })
        }

        @JvmStatic
        fun <T : Comparable<T>>  compareSortedList(
            list:List<T>, compareList: List<T>
        ):ListDiff<T> {
            return compareSortedList(list, compareList, Comparator { o1, o2 -> o1.compareTo(o2) })
        }

        @JvmStatic
        fun <T: Comparable<T>> isListEquals(list: List<T>, compareList: List<T>): Boolean {
            return isListEquals(list, compareList, Comparator { o1, o2 -> o1.compareTo(o2) })
        }

        @JvmStatic
        fun <T: Comparable<T>> isListExactlyEquals(list: List<T>, compareList: List<T>): Boolean {
            return isListExactlyEquals(list, compareList, Comparator { o1, o2 -> o1.compareTo(o2) })
        }

        @JvmStatic
        fun <T> compareListWithoutSorting(
            list: List<T>, compareList: List<T>, comparator: Comparator<T>
        ): ListDiff<T> {
            val removedElements = ArrayList<T>(list)
            val newElements = ArrayList<T>(compareList)
            val sameElements = ArrayList<T>(list.size)

            val iter = removedElements.iterator()
            var cIter: Iterator<T>?

            while (iter.hasNext()) {
                val ele = iter.next()

                cIter = newElements.iterator()

                while (cIter.hasNext()) {
                    val cEle = cIter.next()
                    if (comparator.compare(ele, cEle) == 0) {
                        sameElements.add(ele)
                        cIter.remove()
                        iter.remove()
                        break
                    }
                }
            }

            return ListDiff<T>(newElements, removedElements, sameElements)
        }

        @JvmStatic
        fun <T> compareList(
            list: List<T>, compareList: List<T>, comparator: Comparator<T>
        ): ListDiff<T> {
            val newList = ArrayList<T>(list)
            val newCompareList = ArrayList<T>(compareList)
            Collections.sort(newList, comparator)
            Collections.sort(newCompareList, comparator)

            return compareSortedList(newList, newCompareList, comparator)
        }

        @JvmStatic
        fun <T> compareSortedList(
            list: List<T>, compareList: List<T>, comparator: Comparator<T>
        ): ListDiff<T> {
            val sameElements = ArrayList<T>()
            val newElements = ArrayList<T>(compareList)
            val removedElements = ArrayList<T>(list)

            if (removedElements.isEmpty() || newElements.isEmpty()) {
                return ListDiff(newElements = newElements, removedElements = removedElements, sameElements = sameElements)
            }

            val iter = removedElements.iterator()
            val cIter = newElements.iterator()

            var ele = iter.next()
            var cele = cIter.next()

            while (ele != null && cele != null) {
                val cmp = comparator.compare(ele, cele)

                if (cmp > 0) {
                    if (cIter.hasNext()) {
                        cele = cIter.next()
                    } else {
                        break
                    }
                } else if (cmp < 0) {
                    if (iter.hasNext()) {
                        ele = iter.next()
                    } else {
                        break
                    }
                } else {
                    sameElements.add(ele)
                    iter.remove()
                    cIter.remove()
                    if (iter.hasNext()) {
                        ele = iter.next()
                    } else {
                        break
                    }
                    if (cIter.hasNext()) {
                        cele = cIter.next()
                    } else {
                        break
                    }
                }
            }

            return ListDiff<T>(newElements, removedElements, sameElements)
        }

        @JvmStatic
        fun <T: Comparable<T>> contains(list: List<T>, compareList: List<T>): Boolean {
            val diff = compareList(list, compareList)

            return !diff.hasNewElements()
        }

        @JvmStatic
        fun <T: Comparable<T>> intercept(list: List<T>, compareList: List<T>): Boolean {
            val diff = compareList(list, compareList)

            return diff.hasSameElements()
        }

        @JvmStatic 
        fun <T> contains(list: List<T>, compareList: List<T>, comparator: Comparator<T>): Boolean {
            val diff = compareList(list, compareList, comparator)

            return !diff.hasNewElements()
        }

        @JvmStatic 
        fun <T> intercept(list: List<T>, compareList: List<T>, comparator: Comparator<T>): Boolean {
            val diff = compareList(list, compareList, comparator)

            return diff.hasSameElements()
        }

        @JvmStatic 
        fun <T> isListEquals(list: List<T>, compareList: List<T>, comparator: Comparator<T>): Boolean {
            if (list.size != compareList.size) {
                return false
            } else {
                val newList = ArrayList<T>(list)
                val newCompareList = ArrayList<T>(compareList)

                Collections.sort(newList, comparator)
                Collections.sort(newCompareList, comparator)

                for (i in 0 .. newList.size - 1) {
                    val ele = newList[i]
                    val cele = newCompareList[i]

                    if (comparator.compare(ele, cele) != 0) {
                        return false
                    }
                }

                return true
            }
        }

        @JvmStatic
        fun <T> isListExactlyEquals(list: List<T>, compareList: List<T>, comparator: Comparator<T>): Boolean {
            if (list.size != compareList.size) {
                return false
            } else {
                for (i in 0 .. list.size - 1) {
                    val e1 = list[i]
                    val e2 = compareList[i]

                    if (comparator.compare(e1, e2) != 0) {
                        return false
                    }
                }

                return true
            }
        }

        @JvmStatic 
        fun <T> reorderList(list: List<T>, refList: List<T>) {
            Collections.sort(
                list, fcn@{
                    o1, o2 ->
                    val i1 = refList.indexOf(o1)
                    val i2 = refList.indexOf(o2)
                    if (i1 < i2) {
                        return@fcn - 1
                    } else if (i1 > i2) {
                        return@fcn 1
                    } else {
                        return@fcn 0
                    }
                }
            )
        }

        @JvmStatic
        fun <T> syncList(targetList: MutableList<T>, listDiff: ListDiff<T>): Boolean {
            var modified = false

            if (listDiff.hasRemovedElements()) {
                targetList.removeAll(listDiff.removedElements)
                modified = true
            }

            if (listDiff.hasNewElements()) {
                targetList.addAll(listDiff.newElements)
                modified = true
            }

            return modified
        }

        @JvmStatic 
        fun <T> syncLists(targetList: MutableList<T>, sourceList: List<T>, comparator: Comparator<T>): Boolean {
            return syncList(targetList, compareList(targetList, sourceList, comparator))
        }

        private fun createList(size: Int): List<Int> {
            val list = ArrayList<Int>(size)

            for (i in 0 .. size-1) {
                list.add(MathUtil.randomInteger(0, size))
            }

            return list
        }
    }

    class ListDiff<T>(val newElements: List<T>, val removedElements: List<T>, val sameElements: List<T>) {
        fun hasRemovedElements(): Boolean {
            return removedElements.isNotEmpty()
        }

        fun hasNewElements(): Boolean {
            return newElements.isNotEmpty()
        }

        fun hasSameElements(): Boolean {
            return sameElements.isNotEmpty()
        }

        val isSame: Boolean
            get() = newElements.isEmpty() && removedElements.isEmpty()

        val isNotSame: Boolean
            get() = !isSame

        val newElementCount: Int
            get() = newElements.size

        val removedElementCount: Int
            get() = removedElements.size

        val sameElementCount: Int
            get() = sameElements.size
    }
}